### PR TITLE
Refactor dep installing on startup to only run when needed, and support auto-updating specific optional dependencies

### PR DESCRIPTION
**Pull request by joeyballentine**
This was needed for two reasons: 
1. The "updating deps..." step of the splash screen sometimes took a while, even when it wasn't doing anything.
2. NCNN really should be auto-updated since implementation details change surrounding things added to it. This means that at any point in time an update could break NCNN functionality unless the user manually updates.

### CODE REVIEW

File: src/main/main.ts [hunk 1]
```diff
@@ -341,23 +346,83 @@ const checkPythonEnv = async (splashWindow: BrowserWindowWithSafeIpc) => {
...
-        const pending = requiredDependencies.filter((dep) => {
-            return dep.packages.some((pkg) => {
-                if (installedPackages.has(pkg.packageName)) return false;
-                log.info(`Dependency ${dep.name} (${pkg.packageName}) not found.`);
-                return true;
-            });
+        let requiredIndividualPackages: PyPiPackage[] = [];
+        requiredDependencies.forEach((dependency) => {
+            requiredIndividualPackages = [...requiredIndividualPackages, ...dependency.packages];
+        });
```
Comment (RunDevelopment) [inline src/main/main.ts:367]:
```suggestion
        const requiredIndividualPackages = requiredDependencies.flatMap((dep) => dep.packages);
```

Same for `optionalInvividualPackages` below.

File: src/main/main.ts [hunk 2]
```diff
@@ -341,23 +346,83 @@ const checkPythonEnv = async (splashWindow: BrowserWindowWithSafeIpc) => {
...
+
+        let optionalInvividualPackages: PyPiPackage[] = [];
+        getOptionalDependencies(hasNvidia).forEach((dependency) => {
+            optionalInvividualPackages = [...optionalInvividualPackages, ...dependency.packages];
+        });
+
+        let isInstallingRequired = false;
+        let isUpdating = false;
```
Comment (RunDevelopment) [inline src/main/main.ts:375]: Please move those two further down and make them this:
```suggestion
        const isInstallingRequired = missingRequiredPackages.length > 0);
        const isUpdating = outOfDateRequiredPackages.length > 0 || outOfDateOptionalPackages.length > 0);
```

**Review by RunDevelopment — CHANGES_REQUESTED** [review]

File: src/main/main.ts [hunk 3]
```diff
@@ -576,8 +643,8 @@ const doSplashScreenChecks = async () =>
             await sleep(250);
 
             splash.webContents.send('checking-deps');
-            await checkPythonDeps(splash);
-            await checkNvidiaSmi();
+            const hasNvidia = await checkNvidiaSmi();
```
Comment (RunDevelopment) [inline src/main/main.ts:646]: This check plus its side effects don't need python at all, right? If not, then we could start the promise at the top of the function and only await the result here. That way, the result is likely already ready when we need its result, and we won't have to wait.

File: src/main/main.ts (hunk 2 above)
Comment (joeyballentine) [reply src/main/main.ts:375]: This is why I've missed your reviews :)

File: src/main/main.ts (hunk 3 above)
Comment (joeyballentine) [reply src/main/main.ts:646]: That is correct, we could do that

**Review by RunDevelopment — APPROVED** [review]

**Pull request merged**